source/tilecameragenerator.ts

```diff
@@ -0,0 +1,491 @@
+    }
+
+    /**
+     * Returns the number of tiles along the Y axis
+     * based on the how many of tileSize fit inside the sourceViewport.
+     * @returns - The number of tiles along the Y axis.
+     */
+    public numberOfYTiles(): number {
+        return Math.ceil(this.sourceViewPort[1] / this.tileSize[1]);
+    }
+
+
+    /**
+     * Returns the total number of tiles
+     * based on the how many of tileSize fit inside the sourceViewport.
+     * @returns - The total number of tiles.
+     */
+    public numberOfTiles(): number {
+        return this.numberOfXTiles() * this.numberOfYTiles();
+    }
+
+    /**
+     * Returns the offset of the current tile.
+     * The padding is not included in the offset.
+     * @returns - Current tile offset.
+     */
+    get offset(): [number, number] {
+        return this._currentOffset;
+    }
+
+    /**
+     * Returns the reference to the camera
+     * that has the viewport of the current tile of the sourceCamera.
+     * It returns a default camera if the camera is undefined.
+     * This is the case when the sourceCamera has not been set.
+     * @returns - The reference to the tile viewing camera.
+     */
+    get camera(): Camera {
+        if (this._camera) {
+            return this._camera;
+        } else {
+            return new Camera();
+        }
+    }
+
+    /**
+     * Returns the sourceCamera which viewport should be divided in tiles.
+     * If the sourceCamera has not been set, it returns a default camera.
+     * @returns - The sourceCamera which viewport should be divided in tiles.
+     */
+    get sourceCamera(): Camera {
+        if (this._sourceCamera) {
+            return this._sourceCamera;
+        } else {
+            return new Camera();
+        }
+    }
+
+    /**
+     * Sets the sourceCamera which viewport should be divided in tiles.
+     * Additionally it creates a deep copy of the sourceCamera
+     * which is used as the tiled camera.
+     * @param camera - The sourceCamera which viewport should be divided in tiles.
+     */
+    set sourceCamera(camera: Camera) {
+        if (this._sourceCamera !== camera) {
+            this._sourceCamera = camera;
+            this._camera = camera.copy();
+            this._valid = false;
+        }
+    }
+
+    /**
+     * Returns the current tile index.
+     * @returns - The current tile index.
+     */
+    get tile(): number {
+        return this._tile;
+    }
+
+    /**
+     * Sets the current tile index.
+     * @param index - The new index.
+     */
+    set tile(index: number) {
+        if (this._tile !== index) {
+            this._tile = index;
+            this._valid = false;
+        }
+    }
+
+    /**
+     * Returns the size of the original viewport
+     * which should be divided in tiles based on the tile size.
+     * If the viewport has not been set, it returns [-1, -1].
+     * @returns - Size of the Viewport.
+     */
+    get sourceViewPort(): [number, number] {
```

Comment:
We use *viewport* as a single word. The correct camel-case name would be `sourceViewport`.